### PR TITLE
prevent debugpy from getting stuck

### PR DESCRIPTION
**Pull request by DavidKutu**
-prevent debugpy from getting stuck by running code
-if we can't continue in RBL, disconnect

For #7586, #7584 and parts 1 and 7 of #7218

-   [ ] Pull request represents a single change (i.e. not fixing disparate/unrelated things in a single PR).
-   [ ] Title summarizes what is changing.
-   [ ] Has a [news entry](https://github.com/Microsoft/vscode-jupyter/tree/main/news) file (remember to thank yourself!).
-   [ ] Appropriate comments and documentation strings in the code.
-   [ ] Has sufficient logging.
-   [ ] Has telemetry for enhancements.
-   [ ] Unit tests & system/integration tests are added/updated.
-   [ ] [Test plan](https://github.com/Microsoft/vscode-jupyter/blob/main/.github/test_plan.md) is updated as appropriate.
-   [ ] [`package-lock.json`](https://github.com/Microsoft/vscode-jupyter/blob/main/package-lock.json) has been regenerated by running `npm install` (if dependencies have changed).

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -865,12 +866,17 @@
...
+                    "command": "jupyter.runAndDebugCell",
+                    "when": "notebookType == jupyter-notebook && jupyter.ispythonnotebook && notebookCellType == code && isWorkspaceTrusted && !config.notebook.consolidatedRunButton",
+                    "group": "inline/cell@1"
                 }
             ],
             "notebook/cell/execute": [
                 {
                     "command": "jupyter.runAndDebugCell",
-                    "when": "notebookType == jupyter-notebook && jupyter.ispythonnotebook && notebookCellType == code && isWorkspaceTrusted && !jupyter.notebookeditor.runByLineInProgress"
+                    "when": "notebookType == jupyter-notebook && jupyter.ispythonnotebook && notebookCellType == code && isWorkspaceTrusted && config.notebook.consolidatedRunButton"
```
Comment (roblourens) [inline package.json:879]: Why this change? It always goes in that menu regardless of the setting by design. Probably will rename the setting or rethink this but I don't want it on the cell toolbar.

File: src/client/debugger/jupyter/debuggingManager.ts [hunk 2]
```diff
@@ -92,102 +93,119 @@ export class DebuggingManager implements IExtensionSingleActivationService, IDeb
...
-                    const editor = this.vscNotebook.activeNotebookEditor;
-                    const range = editor?.selections[0];
-                    if (range) {
-                        cell = editor?.document.cellAt(range.start);
+                }, 1000)
+            ),
+
+            this.commandManager.registerCommand(
+                DSCommands.RunByLineNext,
+                throttle((cell: NotebookCell | undefined) => {
```
Comment (roblourens) [inline src/client/debugger/jupyter/debuggingManager.ts:145]: Why throttle this? When I am doing normal javascript debugging, I can step as fast as I want. I can even hold down F10 to step really fast. The same should work here.

File: src/client/debugger/jupyter/debuggingManager.ts [hunk 3]
```diff
@@ -92,102 +93,119 @@ export class DebuggingManager implements IExtensionSingleActivationService, IDeb
...
-            this.commandManager.registerCommand(DSCommands.RunAndDebugCell, async (cell: NotebookCell | undefined) => {
-                sendTelemetryEvent(DebuggingTelemetry.clickedRunAndDebugCell);
-                const editor = this.vscNotebook.activeNotebookEditor;
-                if (!cell) {
-                    const range = editor?.selections[0];
-                    if (range) {
-                        cell = editor?.document.cellAt(range.start);
+            this.commandManager.registerCommand(
+                DSCommands.RunAndDebugCell,
+                throttle(async (cell: NotebookCell | undefined) => {
```
Comment (roblourens) [inline src/client/debugger/jupyter/debuggingManager.ts:183]: I worry it will feel broken if this command sometimes doesn't work if stuff happens too quickly

Comment (rchiodo) [reply src/client/debugger/jupyter/debuggingManager.ts:183]: Yeah throttle of 1 second seems like way too much. Maybe like 100ms

File: package.json (hunk 1 above)
Comment (DavidKutu) [reply package.json:879]: why not? 

I moved it because some people will probably want the setting to get rid of the drop down on the run button

Comment (roblourens) [reply package.json:879]: I'm not going to add another button to every cell toolbar unless masses of angry users kick down my door

File: src/client/debugger/jupyter/debuggingManager.ts (hunk 2 above)
Comment (DavidKutu) [reply src/client/debugger/jupyter/debuggingManager.ts:145]: Because that seems to be the issue, clicking fast enough causes debugpy to freeze.

The changes to the controller are what gets it out of the situation (finishes that frozen session and un freezes debugpy).
The throttling is meant to prevent the situation.

Its removable, but I'd rather have at least some.

Comment (roblourens) [reply src/client/debugger/jupyter/debuggingManager.ts:145]: It's hard to believe that debugpy just can't step fast. I can debug a normal python script and step over 50 lines in a couple seconds

Comment (DonJayamanne) [reply src/client/debugger/jupyter/debuggingManager.ts:145]: I agree, I think we need to identify the root cause, this feels like a bandaid

File: src/client/debugger/jupyter/debugControllers.ts [hunk 4]
```diff
@@ -57,13 +57,15 @@ export class RunByLineController implements IDebuggingDelegate {
...
             traceVerbose(`No paused thread, can't do RBL`);
+            this.stop();
             return;
         }
 
         void this.debugAdapter.stepIn(this.lastPausedThreadId);
     }
 
     public stop(): void {
+        void executeDebugPyFix(this.kernel);
```
Comment (DonJayamanne) [inline src/client/debugger/jupyter/debugControllers.ts:68]: Since comments would be good here, didn't make sense why this is required

File: src/client/debugger/jupyter/debugControllers.ts [hunk 5]
```diff
@@ -171,12 +173,15 @@ async function cellDebugSetup(
...
     debugCell: NotebookCell
 ): Promise<void> {
+    await executeDebugPyFix(kernel);
+    await debugAdapter.dumpCell(debugCell.index);
+}
+
+async function executeDebugPyFix(kernel: IKernel) {
     // remove this if when https://github.com/microsoft/debugpy/issues/706 is fixed and ipykernel ships it
     // executing this code restarts debugpy and fixes https://github.com/microsoft/vscode-jupyter/issues/7251
     if (kernel) {
```
Comment (DonJayamanne) [inline src/client/debugger/jupyter/debugControllers.ts:183]: This is always not undefined

File: src/client/debugger/jupyter/debuggingManager.ts [hunk 6]
```diff
@@ -92,102 +93,119 @@ export class DebuggingManager implements IExtensionSingleActivationService, IDeb
...
-            this.commandManager.registerCommand(DSCommands.DebugNotebook, async () => {
-                const editor = this.vscNotebook.activeNotebookEditor;
-                if (editor) {
-                    if (await this.checkForIpykernel6(editor.document)) {
-                        this.updateToolbar(true);
-                        void this.startDebugging(editor.document);
+            this.commandManager.registerCommand(
+                DSCommands.DebugNotebook,
+                throttle(async () => {
+                    const editor = this.vscNotebook.activeNotebookEditor;
```
Comment (DonJayamanne) [inline src/client/debugger/jupyter/debuggingManager.ts:99]: I think the solution is, if we're already debugging or have started then return from here.
I'd if in progress then get out, mashed three throttle unnecessary.

File: src/client/debugger/jupyter/debuggingManager.ts [hunk 7]
```diff
@@ -92,102 +93,119 @@ export class DebuggingManager implements IExtensionSingleActivationService, IDeb
...
+                    if (!cell) {
+                        return;
+                    }
 
-                if (editor) {
-                    if (await this.checkForIpykernel6(editor.document, DataScience.startingRunByLine())) {
-                        this.updateToolbar(true);
-                        this.updateCellToolbar(true);
-                        await this.startDebuggingCell(editor.document, KernelDebugMode.RunByLine, cell);
+                    if (editor) {
```
Comment (DonJayamanne) [inline src/client/debugger/jupyter/debuggingManager.ts:129]: Same comment as earlier

File: src/client/debugger/jupyter/debuggingManager.ts (hunk 2 above)
Comment (DonJayamanne) [reply src/client/debugger/jupyter/debuggingManager.ts:145]: I think they problem is not because it happens fast, it's because we start multiple debug sessions instead of checkingbif we have already started one

**Review by DonJayamanne — CHANGES_REQUESTED** [review]
We don't need the throttle

File: src/client/debugger/jupyter/debugControllers.ts (hunk 4 above)
Comment (roblourens) [reply src/client/debugger/jupyter/debugControllers.ts:68]: I'm also curious about this, we execute this code when we start, why do we have to execute it when we stop?

File: src/client/debugger/jupyter/debuggingManager.ts [hunk 8]
```diff
@@ -95,6 +95,9 @@ export class DebuggingManager implements IExtensionSingleActivationService, IDeb
             this.commandManager.registerCommand(DSCommands.DebugNotebook, async () => {
                 const editor = this.vscNotebook.activeNotebookEditor;
                 if (editor) {
+                    if (this.notebookToDebugger.has(editor.document)) {
```
Comment (DonJayamanne) [inline src/client/debugger/jupyter/debuggingManager.ts:98]: Not sure this is going to be enough.
If user clicks this button while `checkforIpykernel6` is still running, then yo'ull end up with duplicate sessions.

Comment (DonJayamanne) [reply src/client/debugger/jupyter/debuggingManager.ts:98]: I think you can move this entire if condition from line 98 to 109 into a seprate fucntion, this has been repeated in 3 places, and then you can easily check if that method has already been called once before if `yes`, then return.

File: src/client/debugger/jupyter/debuggingManager.ts [hunk 9]
```diff
@@ -232,6 +212,50 @@ export class DebuggingManager implements IExtensionSingleActivationService, IDeb
         this.runByLineInProgress.set(runningByLine).ignoreErrors();
     }
 
+    private async tryToStartDebugging(mode: KernelDebugMode, editor?: NotebookEditor, cell?: NotebookCell) {
+        if (editor) {
+            if (this.notebookInProgress.has(editor.document)) {
```
Comment (roblourens) [inline src/client/debugger/jupyter/debuggingManager.ts:217]: This is checking the case where debugging is starting, but doesn't it still need to check `this.notebookToDebugger.has`?

File: src/client/debugger/jupyter/debuggingManager.ts [hunk 10]
```diff
@@ -243,6 +208,50 @@ export class DebuggingManager implements IExtensionSingleActivationService, IDeb
...
+                return;
```
Comment (DonJayamanne) [inline src/client/debugger/jupyter/debuggingManager.ts:218]: I think returning early such as 

```
if (!editor) {
	return;
}
```

& a few of thse will improve readabiliyty of this method, right now its nested way too much.

File: src/client/debugger/jupyter/debuggingManager.ts [hunk 11]
```diff
@@ -243,6 +208,50 @@ export class DebuggingManager implements IExtensionSingleActivationService, IDeb
...
+                            if (cell) {
+                                this.updateToolbar(true);
+                                void this.startDebuggingCell(editor.document, KernelDebugMode.Cell, cell);
+                            }
+                            break;
+                        case KernelDebugMode.RunByLine:
+                            if (cell) {
+                                this.updateToolbar(true);
+                                this.updateCellToolbar(true);
+                                await this.startDebuggingCell(editor.document, KernelDebugMode.RunByLine, cell);
```
Comment (DonJayamanne) [inline src/client/debugger/jupyter/debuggingManager.ts:244]: in line 233 we're not awaiting, but here we're awaiting.
Confused why that's the case, comments would help.

Comment (DavidKutu) [reply src/client/debugger/jupyter/debuggingManager.ts:244]: It'll be good to await them all

File: src/client/debugger/jupyter/debuggingManager.ts [hunk 12]
```diff
@@ -232,6 +212,51 @@ export class DebuggingManager implements IExtensionSingleActivationService, IDeb
...
+            ) {
+                switch (mode) {
+                    case KernelDebugMode.Everything:
+                        this.updateToolbar(true);
+                        await this.startDebugging(editor.document);
+                        break;
+                    case KernelDebugMode.Cell:
+                        if (cell) {
+                            this.updateToolbar(true);
+                            await this.startDebuggingCell(editor.document, KernelDebugMode.Cell, cell);
```
Comment (DonJayamanne) [inline src/client/debugger/jupyter/debuggingManager.ts:241]: I'm sorry, but I'm not convinced we need `awaits` here
The old code didn't have awaits, feels like there's something we're not understanding here 
I.e. why do we need an await or what happens if there is no await?

Would like to have some comments why we must await, e.g. tomorrow what would stop working if we removed it.

**Comment by roblourens** [comment]
It took longer (this is after many attempts) but I can still break it

Sep-22-2021 18-37-12

**Comment by DavidKutu** [comment]
@roblourens, the latest commit should catch your scenario

File: src/client/debugger/jupyter/debuggingManager.ts [hunk 13]
```diff
@@ -218,7 +218,15 @@ export class DebuggingManager implements IExtensionSingleActivationService, IDeb
             return;
         }
 
-        if (this.notebookInProgress.has(editor.document) || this.isDebugging(editor.document)) {
+        if (this.notebookInProgress.has(editor.document)) {
+            return;
+        }
+
+        if (this.isDebugging(editor.document)) {
+            this.updateToolbar(true);
```
Comment (roblourens) [inline src/client/debugger/jupyter/debuggingManager.ts:226]: Why do you have to do this here? If you are already debugging, shouldn't it already be in this state?

Comment (DavidKutu) [reply src/client/debugger/jupyter/debuggingManager.ts:226]: it should, but that's the buggy state you get trapped in.

this would correct the context keys

Comment (DonJayamanne) [reply src/client/debugger/jupyter/debuggingManager.ts:226]: Not sure i like this, feels like we're re-executing some code & that seems to fix a problem (but we don't know why).
I would like us to fix the root cause, why is this required in the first place

Comment (DavidKutu) [reply src/client/debugger/jupyter/debuggingManager.ts:226]: The root cause is out of our hands, its debugpy freezing inside ipykernel.

And I disagree, we should be trying to fix root causes, but since the root cause is not on our code, we should also try to fix it if it happens.

File: src/client/debugger/jupyter/debuggingManager.ts (hunk 12 above)
Comment (DavidKutu) [reply src/client/debugger/jupyter/debuggingManager.ts:248]: it helps with keeping the context keys aligned with the status of the debugger. My most recent commit puts them after the awaits.

File: src/client/debugger/jupyter/debuggingManager.ts (hunk 13 above)
Comment (DonJayamanne) [reply src/client/debugger/jupyter/debuggingManager.ts:226]: > its debugpy freezing inside ipykernel.

If that's the cause, how can updating the toolbar fix anything in the kernel (i.e. updating a UI will not address anything in ipykernel freezing).

Comment (roblourens) [reply src/client/debugger/jupyter/debuggingManager.ts:226]: It seems like the root cause is us having the context keys out of sync with the RBL lifecycle. Your latest commit is a good one towards fixing that, I think previously we reset the context keys too early, before debugging actually stopped.

Fundamentally, at this point in the code, two things are out of sync. It's hard to point at one and say this one is always right. It should be possible to write the code so that they are always in sync.

**Review by DonJayamanne — COMMENTED** [review]
.

**Review by DonJayamanne — APPROVED** [review]
Removing the unblocking request.

**Review by DonJayamanne — COMMENTED** [review]
.

File: src/client/debugger/jupyter/debuggingManager.ts [hunk 14]
```diff
@@ -241,19 +241,25 @@ export class DebuggingManager implements IExtensionSingleActivationService, IDeb
...
-                        this.updateToolbar(true);
+                        if (this.isDebugging(editor.document)) {
+                            this.updateToolbar(true);
+                        }
                         break;
                     case KernelDebugMode.Cell:
                         if (cell) {
                             await this.startDebuggingCell(editor.document, KernelDebugMode.Cell, cell);
-                            this.updateToolbar(true);
+                            if (this.isDebugging(editor.document)) {
```
Comment (DonJayamanne) [inline src/client/debugger/jupyter/debuggingManager.ts:251]: Under what circustances will this `if` condition be `false`, i don't think this can ever be `false`, its aleways `true`, it should be true, else the previous await should throw an error.
Having this `if` makes it look as though the previous code can run to completion without errors & not start the debugger (which i think is impossible)

File: src/client/debugger/jupyter/debuggingManager.ts [hunk 15]
```diff
@@ -241,19 +241,25 @@ export class DebuggingManager implements IExtensionSingleActivationService, IDeb
...
+                                this.updateToolbar(true);
+                            }
                         }
                         break;
                     case KernelDebugMode.RunByLine:
                         if (cell) {
                             await this.startDebuggingCell(editor.document, KernelDebugMode.RunByLine, cell);
-                            this.updateToolbar(true);
-                            this.updateCellToolbar(true);
+                            if (this.isDebugging(editor.document)) {
```
Comment (DonJayamanne) [inline src/client/debugger/jupyter/debuggingManager.ts:259]: As earlier, i don't think this if condition is required.

File: src/client/debugger/jupyter/debuggingManager.ts (hunk 14 above)
Comment (DavidKutu) [reply src/client/debugger/jupyter/debuggingManager.ts:251]: I guess that's true, it just feels safer.
I'll remove them

File: src/client/debugger/jupyter/debugControllers.ts [hunk 16]
```diff
@@ -57,13 +57,16 @@ export class RunByLineController implements IDebuggingDelegate {
...
+            this.stop();
             return;
         }
 
         void this.debugAdapter.stepIn(this.lastPausedThreadId);
     }
 
     public stop(): void {
+        // When debugpy gets stuck, running a cell fixes it and allows us to start another debugging session
+        void this.kernel.executeHidden('pass');
```
Comment (DonJayamanne) [inline src/client/debugger/jupyter/debugControllers.ts:69]: Isn't this a hack, i think we should get to the bottom of this.

Comment (DavidKutu) [reply src/client/debugger/jupyter/debugControllers.ts:69]: I'll file a bug in ipykernel, and mention it here, but I don't see what else we can do

File: src/client/debugger/jupyter/debugControllers.ts [hunk 17]
```diff
@@ -57,13 +57,16 @@ export class RunByLineController implements IDebuggingDelegate {
     public continue(): void {
         if (typeof this.lastPausedThreadId !== 'number') {
             traceVerbose(`No paused thread, can't do RBL`);
+            this.stop();
```
Comment (roblourens) [inline src/client/debugger/jupyter/debugControllers.ts:60]: With this, if I click the RBL button and then click it again before the first pause, it disconnects. That's not that bad of a problem but could be annoying, especially since it will then run the whole cell, even if I didn't want/expect it to. 

I guess this is working around a case where we attach to debugpy and it doesn't respond, but if we have fixed the issue on our end of trying to attach twice at a time, then I don't understand how else we get into that case

File: src/client/debugger/jupyter/debugControllers.ts (hunk 16 above)
Comment (roblourens) [reply src/client/debugger/jupyter/debugControllers.ts:69]: > When debugpy gets stuck

It would be good to explain how that happens

**Comment by roblourens** [comment]
Played with it for awhile and was not able to totally break it, nice!
